### PR TITLE
Refactor API to replace DocParticle with DocExcerpt

### PR DESCRIPTION
**Pull request by pgonzal**
This PR simplifies the TSDoc AST structure based on various hindsights from using the library:

- The `DocParticle` nodes are only useful when the tree is constructed from a parsed input. For manually constructed nodes, it's wasteful to construct them.  To reflect this change, `DocParticle` has been renamed to `DocExcerpt`, and the old `Excerpt` helper was removed.
- All excerpts should be represented in a uniform way by a single class.  It's confusing for some `DocNode` subclasses to represent excerpts.  To reflect this change, `DocNodeLeaf` has been eliminated.
- `DocNode.updateParameters()` was awkward and really only needed for the case where a `DocInlineTag` is promoted to a `DocLinkTag`, which also had a weirdness where the `DocInlineTag` base class had a "content" excerpt that overlapped with the `DocLinkTag` child class's specialized structure.  To implement this change, `DocInlineTag` and `DocLinkTag` now inherit from a new abstract base class `DocInlineTagBase`, and `DocNode.updateParameters()`  is removed.
- The old `Excerpt` class had an `Excerpt.spacingAfterContent` property that had a couple problems (1) often it was unused, making it unclear which node a space was associated with, and (2) it could not represent situations where spacing precedes the content.  With this PR, spacing is no longer modeled specially, and instead is simply another `DocExcerpt.kind`.
- `TrimSpacesTransform` no longer tries to merge/remap excerpts when collapsing `DocPlainText` nodes.  The semantics were weird, and nobody was using that feature.  If we want to bring it back, instead we should track the `DocNode` objects that were absorbed by the transform.

### CODE REVIEW

File: tsdoc/src/nodes/DocCodeSpan.ts [hunk 1]
```diff
@@ -23,61 +29,56 @@ export class DocCodeSpan extends DocNode {
...
-  public constructor(parameters: IDocCodeSpanParameters) {
+  public constructor(parameters: IDocCodeSpanParameters | IDocCodeSpanParsedParameters) {
     super(parameters);
+
+    if (DocNode.isParsedParameters(parameters)) {
+      this._openingDelimiterExcerpt = new DocExcerpt({
+        excerptId: ExcerptId.CodeSpan_OpeningDelimiter,
+        content: parameters.openingDelimiterExcerpt
+      });
+      this._openingDelimiterExcerpt = new DocExcerpt({
```
Comment (iclanton) [inline tsdoc/src/nodes/DocCodeSpan.ts]: >_openingDelimiterExcerpt [](start = 11, length = 24)

Copy + paste error.

**Comment by iclanton** [comment]
> export class DocParamBlock extends DocBlock {

This shouldn't inherit from DocContainer, it should contain a DocContainer.

---
Refers to: tsdoc/src/nodes/DocParamBlock.ts:32 in d2b39fe. [](commit_id = d2b39feef843ac00c8eab8eeb2c8de717d7b88c3, deletion_comment = False)

File: tsdoc/src/nodes/DocErrorText.ts [hunk 2]
```diff
@@ -15,35 +16,56 @@ export interface IDocErrorTextParameters extends IDocNodeLeafParameters {
...
+
+  private readonly _errorMessage: string;
+  private readonly _errorLocation: TokenSequence;
 
   /**
    * Don't call this directly.  Instead use {@link TSDocParser}
    * @internal
    */
-  public constructor(parameters: IDocErrorTextParameters) {
+  public constructor(parameters: IDocErrorTextParsedParameters) {
```
Comment (iclanton) [inline tsdoc/src/nodes/DocErrorText.ts]: >constructor [](start = 9, length = 11)

Throw if non-parsed?

File: tsdoc/src/nodes/DocLinkTag.ts [hunk 3]
```diff
@@ -1,46 +1,116 @@
...
+
+    if (this.tagNameWithUpperCase !== '@LINK') {
+      throw new Error('DocLinkTag requires the tag name to be "{@link}"');
+    }
+
+    this._codeDestination = parameters.codeDestination;
+
+    if (DocNode.isParsedParameters(parameters)) {
+      if (parameters.codeDestination !== undefined) {
+        if (parameters.urlDestinationExcerpt !== undefined) {
```
Comment (iclanton) [inline tsdoc/src/nodes/DocLinkTag.ts]: Collapse these.

File: tsdoc/src/nodes/DocLinkTag.ts [hunk 4]
```diff
@@ -1,46 +1,116 @@
...
+
+      if (parameters.linkTextExcerpt) {
+        this._linkTextExcerpt = new DocExcerpt({
+          excerptId: ExcerptId.LinkTag_LinkText,
+          content: parameters.linkTextExcerpt
+        });
+      }
+    } else {
+      if (parameters.codeDestination !== undefined) {
+        if (parameters.urlDestination !== undefined) {
```
Comment (iclanton) [inline tsdoc/src/nodes/DocLinkTag.ts]: Collapse these.

File: tsdoc/src/nodes/DocLinkTag.ts [hunk 5]
```diff
@@ -1,46 +1,116 @@
...
+          throw new Error('Either the codeDestination or the urlDestination may be specified, but not both');
+        }
+      }
+
+      this._urlDestination = parameters.urlDestination;
```
Comment (iclanton) [inline tsdoc/src/nodes/DocLinkTag.ts]: >_urlDestination [](start = 11, length = 15)

Move this up to next to _codeDestination.

File: tsdoc/src/nodes/DocExcerpt.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  PlainText = 'PlainText',
+
+  SoftBreak = 'SoftBreak'
+}
+
+/**
+ * Constructor parameters for {@link DocExcerpt}.
+ */
+export interface IDocExcerptParameters extends IDocNodeParameters {
```
[the remark this answers is not in the document]
Comment (pgonzal) [reply tsdoc/src/nodes/DocExcerpt.ts]: I tried this, but it adds a ton of `parsed: true` lines throughout the code base that don't serve any real purpose.

---
In reply to: [225368272](https://github.com/Microsoft/tsdoc/pull/108#discussion_r225368272) [](ancestors = 225368272)

**Comment by pgonzal** [comment]
> export class DocParamBlock extends DocBlock {

This is a good improvement, but I'm going to handle it in a separate PR, because it's a significant change

---
In reply to: [430071292](https://github.com/Microsoft/tsdoc/pull/108#issuecomment-430071292) [](ancestors = 430071292)

---
Refers to: tsdoc/src/nodes/DocParamBlock.ts:32 in d2b39fe. [](commit_id = d2b39feef843ac00c8eab8eeb2c8de717d7b88c3, deletion_comment = False)

**Pull request merged**